sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_launcher.py

```diff
@@ -0,0 +1,199 @@
+                      help='Query to run')
+
+  parser.add_argument('--subscription',
+                      type=str,
+                      help='Pub/Sub subscription to read from')
+
+  parser.add_argument('--topic',
+                      type=str,
+                      help='Pub/Sub topic to read from')
+
+  args, pipeline_args = parser.parse_known_args()
+  options = PipelineOptions(pipeline_args)
+  print('args, pipeline_args:', args, pipeline_args)
+
+  # Usage with Dataflow requires a project to be supplied.
+  if options.view_as(GoogleCloudOptions).project is None:
+    parser.print_usage()
+    print(sys.argv[0] + ': error: argument --project is required')
+    sys.exit(1)
+
+  # Pub/Sub is currently available for use only in streaming pipelines.
+  if options.view_as(StandardOptions).streaming is None:
+    parser.print_usage()
+    print(sys.argv[0] + ': error: argument --streaming is required')
+    sys.exit(1)
+
+  # wait_until_finish ensures that the streaming job is canceled.
+  if options.view_as(TestOptions).wait_until_finish_duration is None:
+    parser.print_usage()
+    print(sys.argv[0] + ': error: argument --wait_until_finish_duration is required') # pylint: disable=line-too-long
+    sys.exit(1)
+
+  # We use the save_main_session option because one or more DoFn's in this
+  # workflow rely on global context (e.g., a module imported at module level).
+  options.view_as(SetupOptions).save_main_session = True
+  return args, options
+
+
+def generate_events(pipeline, args, pipeline_options):
+  from google.cloud import pubsub
+  publish_client = pubsub.Client(project=PUBSUB_PROJECT)
+  topic = publish_client.topic(PUBSUB_TOPIC)
+  sub = topic.subscription(PUBSUB_SUBSCRIPTION)
+  assert topic.exists()
+  assert sub.exists()
+
+  logging.info('Generating %d auction events to topic %s',
+               EVENT_COUNT_PER_TYPE*3, topic.name)
+  for _ in xrange(EVENT_COUNT_PER_TYPE):
+    topic.publish(PERSON_EVENT)
+    topic.publish(AUCTION_EVENT)
+    topic.publish(BID_EVENT)
+
+  logging.info('Finished event generation.')
+
+  # Read from PubSub into a PCollection.
+  if args.subscription:
+    raw_events = pipeline | 'ReadPubSub' >> beam.io.ReadFromPubSub(
+        subscription=args.subscription)
+  else:
+    raw_events = pipeline | 'ReadPubSub' >> beam.io.ReadFromPubSub(
+        topic=args.topic)
+
+  return raw_events
+
+
+def run_query(query, args, pipeline_options):
+  try:
+    pipeline = beam.Pipeline(options=pipeline_options)
+    raw_events = generate_events(pipeline, args, pipeline_options)
+    query.load(raw_events)
+    result = pipeline.run()
+    job_duration = pipeline_options.view_as(TestOptions).wait_until_finish_duration # pylint: disable=line-too-long
+    if pipeline_options.view_as(StandardOptions).runner == 'DataflowRunner':
+      result.wait_until_finish(duration=job_duration)
+      result.cancel()
+    else:
+      result.wait_until_finish()
+  except NotImplementedError as exp:
```

Comment:
What would throw NotImplementedError in the above code?

Reply:
Oh, this was a slip for when I was trying to use the result of wait_until_finish in combination with the DirectRunner (which doesn't implement that method). It is not needed now, so I'll remove it.